joblib/numpy_pickle_utils.py

```diff
@@ -0,0 +1,589 @@
+"""Utilities for fast persistence of big data, with optional compression."""
+
+# Author: Gael Varoquaux <gael dot varoquaux at normalesup dot org>
+# Copyright (c) 2009 Gael Varoquaux
+# License: BSD Style, 3 clauses.
+
+import pickle
+import sys
+import io
+import zlib
+import bz2
+import warnings
+import contextlib
+from contextlib import closing
+
+from ._compat import PY3_OR_LATER, PY26, PY27
+
+try:
+    from threading import RLock
+except ImportError:
+    from dummy_threading import RLock
+
+if PY3_OR_LATER:
+    Unpickler = pickle._Unpickler
+    Pickler = pickle._Pickler
+    xrange = range
+else:
+    Unpickler = pickle.Unpickler
+    Pickler = pickle.Pickler
+
+try:
+    import numpy as np
+except ImportError:
+    np = None
+
+try:
+    import lzma
+except ImportError:
+    lzma = None
+
+
+# Magic numbers of supported compression file formats.        '
+_ZFILE_PREFIX = b'ZF'  # used with pickle files created before 0.9.3.
+_ZLIB_PREFIX = b'\x78'
+_GZIP_PREFIX = b'\x1f\x8b'
+_BZ2_PREFIX = b'BZ'
+_XZ_PREFIX = b'\xfd\x37\x7a\x58\x5a'
+_LZMA_PREFIX = b'\x5d\x00'
+
+# Supported compressors
+_COMPRESSORS = ('zlib', 'bz2', 'lzma', 'xz', 'gzip')
+
+# The max magic number length of supported compression file types.
+_MAX_PREFIX_LEN = max(len(prefix)
+                      for prefix in (_ZFILE_PREFIX, _GZIP_PREFIX, _BZ2_PREFIX,
+                                     _XZ_PREFIX, _LZMA_PREFIX))
+
+# Buffer size used in io.BufferedReader and io.BufferedWriter
+_IO_BUFFER_SIZE = 10 * 1024 ** 2
```

Comment:
Done ! @grisel, I'll now rebase into a single commit and then it should be fine.